src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1870 @@
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //
+  // Small case default: register on entry
+  //
+  //  rbx: haystack
+  //  r14: needle
+  //  r13: k - 1
+  //  r12: k
+  //  r10: n - k
+  //  rbp: -1
+  //  rdi: junk
+  //  rsi: n
+  //  rdx: junk
+  //  rcx: junk
+  //  XMM_BYTE_0 - first element of needle broadcast
+  //  XMM_BYTE_K - last element of needle broadcast
+  //
+  // Set up in broadcast_additional_needles
+  //  XMM_BYTE_1 - second element of needle broadcast
+  //
+  //  Haystack always copied to stack, so 32-byte reads OK
+  //  Haystack length < 32
+  //  10 < needle length < 32
+
+  {
+    __ bind(L_smallCaseDefault);
+
+    Label L_innerLoop;
+
+    const Register firstNeedleCompare = rdx;
+    const Register compLen = r9;
+    const Register haystack = rbx;
+    const Register mask = r8;
+    const Register rTmp = rdi;
+    const Register rTmp2 = r13;
+    const Register rTmp3 = rax;
+
+// r14 and r12 will be re-used later in this procedure
+#undef needle
+#define needle r14
+#undef needleLen
+#define needleLen r12
+
+    broadcast_additional_needles(false, 0 /* unknown */, needle, needleLen, rTmp3, ae, _masm);
+
+    // For small haystacks we already know that the 1st, 2nd, and last bytes of the needle
+    // compare equal, so we can reduce the byte count to arrays_equals
+    __ leaq(firstNeedleCompare, Address(needle, isU ? 4 : 2));
+    __ leaq(compLen, Address(needleLen, isU ? -6 : -3));
+
+    //  firstNeedleCompare has address of third element of needle
+    //  compLen has length of comparison to do (3 elements less than needle size)
+
+    // Helper to compare the 1st, 2nd, and last byte of the needle to the haystack
+    // in the correct position.  Since the haystack is < 32 bytes, not finding matching
+    // needle bytes can just return failure.  Otherwise, we loop through the found
+    // matches.
+    compare_haystack_to_needle(false, 0, L_returnError, haystack, mask, needleLen, rTmp3, XMM_TMP1,
+                               XMM_TMP2, ae, _masm);
+
+// NOTE: REGISTER RE-USE for r12 and r14
+#undef needle
+#undef saveCompLen
+#define saveCompLen r14
+#undef needleLen
+#undef saveNeedleAddress
+#define saveNeedleAddress r12
+
+    // Save registers stomped by arrays_equals
+    __ movq(saveCompLen, compLen);
+    __ movq(saveNeedleAddress, firstNeedleCompare);  // Save address of 2nd element of needle
+
+    // Find index of a potential match
+    __ align(OptoLoopAlignment);
+    __ bind(L_innerLoop);
+    __ tzcntl(r11, mask);
+
+    __ leaq(rTmp, Address(haystack, r11, Address::times_1, isU ? 4 : 2));
+
+    // Check for needle equality.  Handles UU and LL cases since byte comparison should be exact
+    __C2 arrays_equals(false, rTmp, firstNeedleCompare, compLen, rTmp3, rTmp2, XMM_TMP3, XMM_TMP4,
+                       false /* char */, knoreg);
+    __ testl(rTmp3, rTmp3);
+    __ jne(L_checkRangeAndReturn);
```

Comment:
Why do we have to checkRange here, would it not be always correct? It so we could return r11 directly (by moving into rax).

Reply:
OK